GH-15239: [C++][Parquet] Parquet writer writes decimal as int32/64

wgtmac opened the pull request:
As the parquet [specs](https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#decimal) states, DECIMAL can be used to annotate the following types:
- int32: for 1 <= precision <= 9
- int64: for 1 <= precision <= 18; precision < 10 will produce a warning
- fixed_len_byte_array: precision is limited by the array size. Length n can store <= floor(log_10(2^(8*n - 1) - 1)) base-10 digits
- binary: precision is not limited, but is required. The minimum number of bytes to store the unscaled value should be used.

The aim of this patch is to provide a writer option to use int32 to annotate decimal when 1 <= precision <= 9 and int64 when 10 <= precision <= 18.
* Closes: #15239

wgtmac commented:
@pitrou @emkornfield @wjones127 Could you please take a look? Thanks!

cc @liukun4515

emkornfield commented on cpp/src/parquet/properties.h:
```diff
@@ -199,7 +199,8 @@ class PARQUET_EXPORT WriterProperties {
           data_page_version_(ParquetDataPageVersion::V1),
-          created_by_(DEFAULT_CREATED_BY) {}
+          created_by_(DEFAULT_CREATED_BY),
+          integer_annotate_decimal_(false) {}
```
Another reason to keep the default false, is to not change enable code paths that weren't used before when people update there library version.

wjones127 replied:
I understand the desire for stability, but would like to have a pathway for improved defaults to be released. It's unfortunate we always do a major release while still trying to respect stability, as there's no clear good time to make breaking changes like this.

emkornfield replied:
I agree this is a hard problem, and don't have any guidance.  The one thing I would say is that if Arrow is intended to be developer facing, then IMO developers looking to upgrade versions we should be following the "least surprise" principle as much as possible.  Other opinions are valid here.

pitrou replied:
Hi,
Some belated comments:
* Maybe it's because I am not a native English speaker, but I have a hard time understanding what "integer annotate decimal" means. Is "annotate" a word here?
* I agree that the defaults should remain false for the time being.
* While enabling this does allow for a more compact representation, compression might at least partially alleviate that (though unfortunately FIXED_LEN_BYTE_ARRAY doesn't seem to support any of the delta encodings)

pitrou replied:
Filed https://issues.apache.org/jira/browse/PARQUET-2231 because parquet-mr allows DELTA_BYTE_ARRAY on FIXED_LEN_BYTE_ARRAY, while the spec doesn't say so...

wjones127 reviewed (changes requested):
This generally looks good, but would like to see an additional test for the null code paths.

wgtmac commented on cpp/src/parquet/arrow/arrow_reader_writer_test.cc:
```diff
@@ -4727,5 +4726,74 @@ std::vector<NestedFilterTestCase> GenerateMapFilteredTestCases() {
+
+TYPED_TEST_SUITE(TestIntegerAnnotateDecimalTypeParquetIO, DecimalTestTypes);
+
+TYPED_TEST(TestIntegerAnnotateDecimalTypeParquetIO, SingleDecimalColumn) {
```
Thanks for your review!

I have added the test with null values. It seems that the unsuccessful CI checks are unrelavant to my commit. Please take a look again. Thanks! @wjones127

wjones127 reviewed (changes requested):
I found Decimal256 doesn't work with this code path if `store_schema` is set. Could you also address that? Here is a unit test that should reproduce:

```cpp
TEST(ArrowReadWrite, Decimal256AsInt) {
  using ::arrow::Decimal256;
  using ::arrow::field;

  auto type = ::arrow::decimal256(8, 4);

  const char* json = R"(["1.0000", null, "-1.2345", "-1000.5678",
                         "-9999.9999", "9999.9999"])";
  auto array = ::arrow::ArrayFromJSON(type, json);
  auto table = ::arrow::Table::Make(::arrow::schema({field("root", type)}), {array});

  parquet::WriterProperties::Builder builder;
  // Enforce integer type to annotate decimal type
  auto writer_properties = builder.enable_integer_annotate_decimal()->build();
  auto props_store_schema = ArrowWriterProperties::Builder().store_schema()->build();

  CheckConfiguredRoundtrip(table, table, writer_properties, props_store_schema);
}
```

wjones127 commented on cpp/src/parquet/arrow/schema.cc:
```diff
@@ -354,11 +354,15 @@ Status FieldToNode(const std::string& name, const std::shared_ptr<Field>& field,
       precision = decimal_type.precision();
       scale = decimal_type.scale();
-      length = DecimalType::DecimalSize(precision);
+      if (properties.integer_annotate_decimal() && 1 <= precision && precision <= 18) {
```
Not sure what you mean by that. This PR is how the data is encoded in the buffers, not what logical type is saved in the Parquet file. When read back to Arrow it will always come out as a Decimal, so there's no need for conversion.

wgtmac replied:
> Does c++ arrow support converting int32/int64 to decimal128 and decimal256？

decimal128/decimal256 support construction from integers. But the parquet itself does not use an integer array to represent decimal values.

liukun4515 replied:
Got it, thanks

wgtmac commented on cpp/src/parquet/arrow/arrow_reader_writer_test.cc:
```diff
@@ -4727,5 +4726,84 @@ std::vector<NestedFilterTestCase> GenerateMapFilteredTestCases() {
+class TestIntegerAnnotateDecimalTypeParquetIO : public TestParquetIO<TestType> {
+ public:
+  template <typename ArrayType>
+  void WriteColumn(const std::shared_ptr<ArrayType>& values) {
```
Fixed

wgtmac commented:
> I found Decimal256 doesn't work with this code path if `store_schema` is set. Could you also address that? Here is a unit test that should reproduce:
> 
> ```c++
> TEST(ArrowReadWrite, Decimal256AsInt) {
>   using ::arrow::Decimal256;
>   using ::arrow::field;
> 
>   auto type = ::arrow::decimal256(8, 4);
> 
>   const char* json = R"(["1.0000", null, "-1.2345", "-1000.5678",
>                          "-9999.9999", "9999.9999"])";
>   auto array = ::arrow::ArrayFromJSON(type, json);
>   auto table = ::arrow::Table::Make(::arrow::schema({field("root", type)}), {array});
> 
>   parquet::WriterProperties::Builder builder;
>   // Enforce integer type to annotate decimal type
>   auto writer_properties = builder.enable_integer_annotate_decimal()->build();
>   auto props_store_schema = ArrowWriterProperties::Builder().store_schema()->build();
> 
>   CheckConfiguredRoundtrip(table, table, writer_properties, props_store_schema);
> }
> ```

Fixed and added the test case. Please take a look again. Thanks! @wjones127

pitrou commented on cpp/src/parquet/properties.h:
```diff
@@ -450,6 +451,22 @@ class PARQUET_EXPORT WriterProperties {
       return this->disable_statistics(path->ToDotString());
     }
 
+    /// Enable integer type to annotate decimal type as below:
```
Is "Enable X to do Y" correct English? Just asking, as I'm not a native English speaker, but I would intuitively have expected "Allow X to do Y".

mapleFU commented:
Should we DCHECK the precision when writing arrow? Seems I meet a bug that my precision is wrong, but it write success as Decimal128, and read as int32... @wgtmac

wgtmac commented:
> Should we DCHECK the precision when writing arrow? Seems I meet a bug that my precision is wrong, but it write success as Decimal128, and read as int32... @wgtmac

Could you please provide the process to reproduce it? Or you can directly submit a PR if it is straight-forward.

mapleFU commented:
Sure, I will submit a tiny check tonight
